### PR TITLE
Mark schema fields with `None` default as `Optional` to pass pydantic v2 validation

### PR DESCRIPTION
**Pull request by danielzuegner**
## Summary

This PR marks all `Field`s with `None` default value as `Optional[type]` in the type annotation. This enables to pass the Pydantic validation when deserializing. For instance, the following fails for me with `atomate=0.0.12` and `pydantic=2.5.2`:

```python
from atomate2.common.schemas.elastic import ElasticDocument
ElasticDocument.model_validate_json(ElasticDocument().json())
```
gives:
```
ValidationError: 6 validation errors for ElasticDocument
structure
  Value error, Must provide Structure, the as_dict form, or the proper [type=value_error, input_value=None, input_type=NoneType]
    For further information visit https://errors.pydantic.dev/2.5/v/value_error
elastic_tensor
  Input should be an object [type=model_type, input_value=None, input_type=NoneType]
    For further information visit https://errors.pydantic.dev/2.5/v/model_type
derived_properties
  Input should be an object [type=model_type, input_value=None, input_type=NoneType]
    For further information visit https://errors.pydantic.dev/2.5/v/model_type
fitting_data
  Input should be an object [type=model_type, input_value=None, input_type=NoneType]
    For further information visit https://errors.pydantic.dev/2.5/v/model_type
fitting_method
  Input should be a valid string [type=string_type, input_value=None, input_type=NoneType]
    For further information visit https://errors.pydantic.dev/2.5/v/string_type
order
  Input should be a valid integer [type=int_type, input_value=None, input_type=NoneType]
    For further information visit https://errors.pydantic.dev/2.5/v/int_type
```
This means that unless all fields are not `None` in a `ElasticDocument`, deserializing with validation will fail. This PR fixes the issue. 

The behavior change from `pydantic` v1 to v2 is also confirmed in this [GitHub issue](https://github.com/pydantic/pydantic/issues/4261).

## Checklist

Work-in-progress pull requests are encouraged, but please put [WIP] in the pull request
title.

Before a pull request can be merged, the following items must be checked:

* [ ] Code is in the [standard Python style](https://www.python.org/dev/peps/pep-0008/).
  The easiest way to handle this is to run the following in the **correct sequence** on
  your local machine. Start with running [`ruff`](https://docs.astral.sh/ruff) and `ruff format` on your new code. This will
  automatically reformat your code to PEP8 conventions and fix many linting issues.
* [ ] Doc strings have been added in the [Numpy docstring format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html).
  Run [ruff](https://beta.ruff.rs/docs/rules/#pydocstyle-d) on your code.
* [ ] Type annotations are **highly** encouraged. Run [mypy](http://mypy-lang.org) to
  type check your code.
* [ ] Tests have been added for any new functionality or bug fixes.
* [ ] All linting and tests pass.

Note that the CI system will run all the above checks. But it will be much more
efficient if you already fix most errors prior to submitting the PR. It is highly
recommended that you use the pre-commit hook provided in the repository. Simply run
`pre-commit install` and a check will be run prior to allowing commits.

### CODE REVIEW

File: src/atomate2/common/schemas/phonons.py [hunk 1]
```diff
@@ -307,7 +307,7 @@ def from_forces_born(
             kpath_concrete, npoints=kwargs.get("npoints_band", 101)
         )
 
-        # phonon band structures will always be computed
+        # phonon band structures will always be cmouted
```
Comment (janosh) [inline src/atomate2/common/schemas/phonons.py:310]: this looks like an accidental revert

File: src/atomate2/common/schemas/phonons.py [hunk 2]
```diff
@@ -324,7 +324,9 @@ def from_forces_born(
         new_plotter = PhononBSPlotter(bs=bs_symm_line)
 
         new_plotter.save_plot(
-            "phonon_band_structure.eps", units=kwargs.get("units", "THz")
+            "phonon_band_structure.eps",
+            img_format=kwargs.get("img_format", "eps"),
+            units=kwargs.get("units", "THz"),
```
Comment (janosh) [inline src/atomate2/common/schemas/phonons.py:329]: same here and below

**Review by janosh — CHANGES_REQUESTED** [review]
@danielzuegner Are you able to add a test covering deserialization with optional fields missing?

**Comment by mkhorton** [comment]
> Are you able to add a test covering deserialization with optional fields missing?

@janosh Is this strictly necessary, given that it's pydantic functionality?

**Comment by mkhorton** [comment]
I'm assuming this is related to a change of behavior in Pydantic v2, although I have not been able to find documentation of this yet.

**Comment by janosh** [comment]
> @janosh Is this strictly necessary, given that it's pydantic functionality?

I think it would be nice to have but of course it's really up to @utf. Given that missing `Optional`s cause errors in user land, a test for this seems no less useful than elsewhere.

**Comment by mkhorton** [comment]
To clarify the change this PR makes is that there was an implicit `Optional` previously (the default was `None`, but this was not reflected in the type hint), and this PR makes this an explicit `Optional` instead, so that the default value does not cause a validation error in pydantic v2; in this sense, there is no new functionality here, which is why I'm suggesting a test is not strictly required (the original functionality was also merged without a test, so a fix could also be merged, although of course a test is preferred).

More generally, this kind of functionality is implicitly tested by the validation performed by pydantic itself, which is how this error was detected. Given that this is a regression and the currently released version of atomate2 is broken for elastic workflows (#659) unless this is addressed, I would suggest we prioritize merging or otherwise fixing if this PR is not sufficient. Perhaps there's something I'm missing here, I'm not sure.

My comment here is, in part, because I'm reluctant to add tests for areas I'm not as familiar with, but I would like to see the existing functionality restored.

**Comment by JaGeo** [comment]
@mkhorton While I completely agree that this needs to be urgently fixed, I would like to add that adding tests for this part would be very relevant (e.g., in a future PR). The changes in pydantic disrupted atomate2 quite a lot.

**Comment by janosh** [comment]
> More generally, this kind of functionality is implicitly tested by the validation performed by pydantic itself, which is how this error was detected.

Exactly, which should make the test code minimal. Just putting the 2 lines @danielzuegner posted above in a test function might do the trick and would allow us to catch breaking `pydantic` changes before we make a release in the future.

```py
from atomate2.common.schemas.elastic import ElasticDocument
ElasticDocument.model_validate_json(ElasticDocument().json())
```

**Comment by danielzuegner** [comment]
> > More generally, this kind of functionality is implicitly tested by the validation performed by pydantic itself, which is how this error was detected.
> 
> Exactly, which should make the test code minimal. Just putting the 2 lines @danielzuegner posted above in a test function might do the trick and would allow us to catch breaking `pydantic` changes before we make a release in the future.
> 
> ```python
> from atomate2.common.schemas.elastic import ElasticDocument
> ElasticDocument.model_validate_json(ElasticDocument().json())
> ```

I'm on it, will push something in the next minutes/ hour or so

**Comment by danielzuegner** [comment]
> > > More generally, this kind of functionality is implicitly tested by the validation performed by pydantic itself, which is how this error was detected.
> > 
> > 
> > Exactly, which should make the test code minimal. Just putting the 2 lines @danielzuegner posted above in a test function might do the trick and would allow us to catch breaking `pydantic` changes before we make a release in the future.
> > ```python
> > from atomate2.common.schemas.elastic import ElasticDocument
> > ElasticDocument.model_validate_json(ElasticDocument().json())
> > ```
> 
> I'm on it, will push something in the next minutes/ hour or so

@janosh done, added some tests which passed locally for me on my machine.

**Review by janosh — COMMENTED** [review]
Tests look great! Thanks @danielzuegner!

 I fixed the ruff error about a missing shebang but some tests started failing with

```py
    return PhononBSDOSDoc.from_forces_born(
  File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/atomate2/common/schemas/phonons.py", line 300, in from_forces_born
    kpath_dict, kpath_concrete = cls.get_kpath(
AttributeError: type object 'ModelMetaclass' has no attribute 'get_kpath'
```

Maybe @JaGeo can comment if this has come up before.

**Comment by JaGeo** [comment]
No, it hasn't

File: src/atomate2/common/schemas/phonons.py [hunk 3]
```diff
@@ -190,6 +190,7 @@ class PhononBSDOSDoc(StructureMetadata):
 
     uuids: Optional[PhononUUIDs] = Field("Field including all relevant uuids")
 
+    @classmethod
```
Comment (JaGeo) [inline src/atomate2/common/schemas/phonons.py:193]: Why are there two classmethods?

**Comment by JaGeo** [comment]
get_kpath is a static method. Maybe that is why?

**Review by janosh — APPROVED** [review]
All green here. Thanks again @danielzuegner for this fix and @JaGeo for the pointer on where the CI error came from.

**Comment by JaGeo** [comment]
Thanks, @danielzuegner @mkhorton and @janosh !

**Comment by mkhorton** [comment]
Thanks @danielzuegner, and @janosh for merging!

**Comment by utf** [comment]
Thanks to everyone who helped getting this merged. Glad to see this fixed.

**Pull request merged**